# Connecting to Google Drive

# Web Scraping

In [1]:
import urllib.request
import os
from bs4 import BeautifulSoup
import time

In [2]:
print(os.getcwd())
# os.chdir('...')

c:\Users\Nova\Documents\Code


## Example, back to Dept History

In [3]:
# url = 'https://www.sta.cuhk.edu.hk/about-us/department-history/'
url = 'https://kyalan.github.io/kytest/department-history.html'
req = urllib.request.Request(url)
resp = urllib.request.urlopen(req)

In [4]:
content = resp.read()

# If needed to print or process the content, you can do so here
print(content)

b'\n<!DOCTYPE html>\n<html dir="ltr" lang="en-US" prefix="og: https://ogp.me/ns#">\n\t<head>\n\t\t<meta charset="UTF-8" />\n\t\t<meta http-equiv="X-UA-Compatible" content="IE=edge">\t\t\n\t\t<link rel="pingback" href="https://www.sta.cuhk.edu.hk/xmlrpc.php" />\n\n\t\t<script type="text/javascript" defer>document.documentElement.className = \'js\';</script>\n\t\t<!---<script type="text/javascript" src="magicthumb/magicthumb.js"></script>--->\n\n\t\t<!-- Compressed CSS -->\n\t\t\n<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/foundation-sites@6.6.3/dist/css/foundation.min.css" integrity="sha256-ogmFxjqiTMnZhxCqVmcqTvjfe1Y/ec4WaRj/aQPvn+I=" crossorigin="anonymous" defer>\n\n\t\t<link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/npm/slick-carousel@1.8.1/slick/slick.css" defer/> \n\t\t\n\t\t<!---<link rel="preload" href="https://cdn.jsdelivr.net/npm/foundation-sites@6.6.3/dist/css/foundation.min.css" as="style" integrity="sha256-ogmFxjqiTMnZhxCqVmcqTvjfe1Y/ec4WaRj/aQ

In [5]:
# print(dir(req))
resp.__dict__

{'fp': None,
 'debuglevel': 0,
 '_method': 'GET',
 'headers': <http.client.HTTPMessage at 0x21738387c50>,
 'msg': 'OK',
 'version': 11,
 'status': 200,
 'reason': 'OK',
 'chunked': False,
 'chunk_left': 'UNKNOWN',
 'length': 0,
 'will_close': True,
 'code': 200,
 'url': 'https://kyalan.github.io/kytest/department-history.html'}

In [6]:
# Or directly
resp = urllib.request.urlopen(url)

html = ''
if resp.status==200:   html = resp.read()
print(len(html))

67041


In [7]:
html[:100]

b'\n<!DOCTYPE html>\n<html dir="ltr" lang="en-US" prefix="og: https://ogp.me/ns#">\n\t<head>\n\t\t<meta chars'

In [9]:
type(html)

bytes

In [10]:
soup = BeautifulSoup(html, 'html.parser')
tag_table = soup.table
print(tag_table)

<table border="1" cellpadding="3" cellspacing="0">
<tbody>
<tr>
<th style="text-align: left; background-color: #ccc1d9;" width="23%">Year</th>
<th style="text-align: left; background-color: #ccc1d9;" width="77%">History</th>
</tr>
<tr class="odd">
<td>1979</td>
<td>Statistics (STAT) Programme was set up as an individual programme of study.</td>
</tr>
<tr class="odd">
<td>1981</td>
<td>Establishment of Division of Statistics at Graduate School, CUHK.</td>
</tr>
<tr class="odd">
<td>1982</td>
<td>Foundation of the Department. 17 students were enrolled in the first year with Prof. Howell Tong as the chairman</td>
</tr>
<tr class="odd">
<td>1986</td>
<td>Professor Sik-yum Lee assumed the duty of chairman of the Department.</td>
</tr>
<tr class="odd">
<td>1993</td>
<td>Professor Nai Ng Chan became the third chairman of the Department.</td>
</tr>
<tr class="odd">
<td>1994</td>
<td>The Department office relocated from Science Centre to the current site – Lady Shaw Building.</td>
</tr>
<tr cla

In [11]:
soup.find_all('tr')

[<tr>
 <th style="text-align: left; background-color: #ccc1d9;" width="23%">Year</th>
 <th style="text-align: left; background-color: #ccc1d9;" width="77%">History</th>
 </tr>,
 <tr class="odd">
 <td>1979</td>
 <td>Statistics (STAT) Programme was set up as an individual programme of study.</td>
 </tr>,
 <tr class="odd">
 <td>1981</td>
 <td>Establishment of Division of Statistics at Graduate School, CUHK.</td>
 </tr>,
 <tr class="odd">
 <td>1982</td>
 <td>Foundation of the Department. 17 students were enrolled in the first year with Prof. Howell Tong as the chairman</td>
 </tr>,
 <tr class="odd">
 <td>1986</td>
 <td>Professor Sik-yum Lee assumed the duty of chairman of the Department.</td>
 </tr>,
 <tr class="odd">
 <td>1993</td>
 <td>Professor Nai Ng Chan became the third chairman of the Department.</td>
 </tr>,
 <tr class="odd">
 <td>1994</td>
 <td>The Department office relocated from Science Centre to the current site – Lady Shaw Building.</td>
 </tr>,
 <tr class="odd">
 <td></td>
 <

In [12]:
# Prettify the BeautifulSoup object: pretty_soup
print(tag_table.prettify())

<table border="1" cellpadding="3" cellspacing="0">
 <tbody>
  <tr>
   <th style="text-align: left; background-color: #ccc1d9;" width="23%">
    Year
   </th>
   <th style="text-align: left; background-color: #ccc1d9;" width="77%">
    History
   </th>
  </tr>
  <tr class="odd">
   <td>
    1979
   </td>
   <td>
    Statistics (STAT) Programme was set up as an individual programme of study.
   </td>
  </tr>
  <tr class="odd">
   <td>
    1981
   </td>
   <td>
    Establishment of Division of Statistics at Graduate School, CUHK.
   </td>
  </tr>
  <tr class="odd">
   <td>
    1982
   </td>
   <td>
    Foundation of the Department. 17 students were enrolled in the first year with Prof. Howell Tong as the chairman
   </td>
  </tr>
  <tr class="odd">
   <td>
    1986
   </td>
   <td>
    Professor Sik-yum Lee assumed the duty of chairman of the Department.
   </td>
  </tr>
  <tr class="odd">
   <td>
    1993
   </td>
   <td>
    Professor Nai Ng Chan became the third chairman of the Departm

In [13]:
tag_rows = tag_table.findAll('tr', {'class':'odd'})
print(tag_rows)

[<tr class="odd">
<td>1979</td>
<td>Statistics (STAT) Programme was set up as an individual programme of study.</td>
</tr>, <tr class="odd">
<td>1981</td>
<td>Establishment of Division of Statistics at Graduate School, CUHK.</td>
</tr>, <tr class="odd">
<td>1982</td>
<td>Foundation of the Department. 17 students were enrolled in the first year with Prof. Howell Tong as the chairman</td>
</tr>, <tr class="odd">
<td>1986</td>
<td>Professor Sik-yum Lee assumed the duty of chairman of the Department.</td>
</tr>, <tr class="odd">
<td>1993</td>
<td>Professor Nai Ng Chan became the third chairman of the Department.</td>
</tr>, <tr class="odd">
<td>1994</td>
<td>The Department office relocated from Science Centre to the current site – Lady Shaw Building.</td>
</tr>, <tr class="odd">
<td></td>
<td>Professor Wing-hung Wong served as the chairman of the Department.</td>
</tr>, <tr class="odd">
<td>1998</td>
<td>Professor Yeh Lam succeeded Professor Wing-hong Wong and took up the chairmanship of t

C:\Users\Nova\AppData\Local\Temp\ipykernel_30592\4138549938.py:1: DeprecationWarning: Call to deprecated method findAll. (Replaced by find_all) -- Deprecated since version 4.0.0.
  tag_rows = tag_table.findAll('tr', {'class':'odd'})


We need to take out all words standing for "Professor", including "Prof.".

In [14]:
import re

for tag in tag_rows:
  tag_values = tag.find_all('td')
  year = tag_values[0].text
  message = re.findall(r'Prof(?:essor)?\.?(?:\s[A-Z][a-z-]*)+', tag_values[1].text)
  print(year, message)

1979 []
1981 []
1982 ['Prof. Howell Tong']
1986 ['Professor Sik-yum Lee']
1993 ['Professor Nai Ng Chan']
1994 []
 ['Professor Wing-hung Wong']
1998 ['Professor Yeh Lam', 'Professor Wing-hong Wong']
1999 ['Professor Kim Hung Li']
2000 ['Professor Jian Qing Fan']
 []
 []
2004 ['Professor Ngai Hang Chan']
2009 []
2012 []
2013 ['Professor Qi Man Shao']
2018 ['Professor Xinyuan Song']
2022 []
2024 []
2024 ['Professor Junhui Wang']
2025 []


## Example on Fitness Rooms in LCSD site

![yuenchaukok_gym.png](https://drive.google.com/uc?export=view&id=1451sSsaHw1ugc2b1CRTxqHMjfYYCYxMu)

Gym rooms are provided by the Leisure and Cultural Services Department (LCSD) in the urban area and the New Territories. Let's access into the information site: https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRooms.do

In [16]:
import urllib.request, urllib.parse
from urllib.parse import urljoin
from bs4 import BeautifulSoup
import pandas as pd

In [17]:
url = 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRooms.do'
resp = urllib.request.Request(url=url, headers={'User-Agent': 'Mozilla/5.0'})

html = ''
req = urllib.request.urlopen(resp)
if req.status==200:   html = req.read().decode()
print(f'website retrieved: {len(html)}')

website retrieved: 57604


Note !

```
headers={'User-Agent': 'Mozilla/5.0'}
```
is a must here.


In [18]:
soup = BeautifulSoup(html, 'html.parser')

We aim to obtain all rows of tables - \<tr\> tags and \<td\> for each tag

In [19]:
tags_tr = soup.find_all('tr')
len(tags_tr)

89

In [20]:
tags_td = tags_tr[1].find_all('td')

In [21]:
tags_td

[<td class="venu">Aberdeen Sports Centre</td>,
 <td class="addr">5/F., Aberdeen Municipal Services Building, 203 Aberdeen Main Road, Aberdeen</td>,
 <td class="tel">2555 8909 </td>,
 <td class="detail" width="13%">
 <a href="/clpss/en/webApp/FitnessRoomDetails.do?id=1374" title="Details">Details</a>
 </td>]

In [22]:
rows = []
for tag_tr in tags_tr:
  tags_td = tag_tr.find_all('td')
  row = []
  for tag_td in tags_td:
    value = tag_td.text.strip()
    row.append(value)
  rows.append(row)

In [23]:
rows

[[],
 ['Aberdeen Sports Centre',
  '5/F., Aberdeen Municipal Services Building, 203 Aberdeen Main Road, Aberdeen',
  '2555 8909',
  'Details'],
 ['Aberdeen Tennis and Squash Centre',
  'No. 1, Aberdeen Praya Road, Hong Kong',
  '2553 6130',
  'Details'],
 ['Ap Lei Chau Sports Centre',
  '8 Hung Shing Street, Ap Lei Chau',
  '2554 0832',
  'Details'],
 ['Chai Wan Sports Centre',
  'Yee Shun Street, Chai Wan',
  '2897 9144',
  'Details*'],
 ['Harbour Road Sports Centre',
  '27 Harbour Road, Wan Chai',
  '2827 9684',
  'Details'],
 ['Hong Kong Park Sports Centre',
  '29 Cotton Tree Drive, Central',
  '2521 5072',
  'Details'],
 ['Island East Sports Centre',
  '52 Lei King Road, Sai Wan Ho',
  '2151 4070',
  'Details'],
 ['Java Road Sports Centre',
  '3/F, Java Road Municipal Services Building, 99 Java Road, North Point',
  '2516 9415',
  'Details*'],
 ['Lockhart Road Sports Centre',
  '10/F., Lockhart Road Municipal Services Building, 225 Hennessy Road, Wan Chai',
  '2879 5521',
  'Detail

The link in html may be in form of relative link, says "/clpss/en/webApp/FitnessRoomDetails.do?id=1374" <br>
You can use urljoin with serviceurl = "https://www.lcsd.gov.hk" for making a full link.

In [24]:
serviceurl = 'https://www.lcsd.gov.hk'

dict_tags = []
for tag in tags_tr:
    tags_td = tag.find_all('td')
    if len(tags_td) != 4: continue
    dict_tag = {}
    dict_tag['venu'] = tags_td[0].text
    dict_tag['addr'] = tags_td[1].text
    dict_tag['tel'] = tags_td[2].text
    dict_tag['link'] = urljoin(serviceurl, tags_td[3].a.get('href'))
    dict_tags.append(dict_tag)

In [25]:
dict_tags

[{'venu': 'Aberdeen Sports Centre',
  'addr': '5/F., Aberdeen Municipal Services Building, 203 Aberdeen Main Road, Aberdeen',
  'tel': '2555 8909 ',
  'link': 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRoomDetails.do?id=1374'},
 {'venu': 'Aberdeen Tennis and Squash Centre',
  'addr': 'No. 1, Aberdeen Praya Road, Hong Kong',
  'tel': '2553 6130 ',
  'link': 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRoomDetails.do?id=1375'},
 {'venu': 'Ap Lei Chau Sports Centre ',
  'addr': '8 Hung Shing Street, Ap Lei Chau',
  'tel': '2554 0832',
  'link': 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRoomDetails.do?id=1376'},
 {'venu': 'Chai Wan Sports Centre',
  'addr': 'Yee Shun Street, Chai Wan',
  'tel': '2897 9144 ',
  'link': 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRoomDetails.do?id=1365'},
 {'venu': 'Harbour Road Sports Centre',
  'addr': '27 Harbour Road, Wan Chai',
  'tel': '2827 9684',
  'link': 'https://www.lcsd.gov.hk/clpss/en/webApp/FitnessRoomDetails.do?id=1440'},
 {'

In [26]:
df_url = pd.DataFrame(dict_tags)

In [27]:
df_url

venu  \
0              Aberdeen Sports Centre   
1   Aberdeen Tennis and Squash Centre   
2          Ap Lei Chau Sports Centre    
3              Chai Wan Sports Centre   
4          Harbour Road Sports Centre   
..                                ...   
81               Yau Oi Sports Centre   
82        Yeung Uk Road Sports Centre   
83        Yuen Chau Kok Sports Centre   
84            Yuen Long Sports Centre   
85         Yuen Wo Road Sports Centre   

                                                 addr         tel  \
0   5/F., Aberdeen Municipal Services Building, 20...  2555 8909    
1               No. 1, Aberdeen Praya Road, Hong Kong  2553 6130    
2                    8 Hung Shing Street, Ap Lei Chau   2554 0832   
3                           Yee Shun Street, Chai Wan  2897 9144    
4                           27 Harbour Road, Wan Chai   2827 9684   
..                                                ...         ...   
81                             Hing On Lane, Tuen Mun  2450 8850    
82  4/F., Yeung Uk Road Municipal Services Buildin...  2415 4445    
83                35 Ngan Shing Street, Sha Tin, N.T.   2509 9108   
84  3/F, Yuen Long Leisure and Cultural Building, ...   2891 9207   
85                            8 Yuen Wo Road, Sha Tin  2604 5987    

                                                 link  
0   https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
1   https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
2   https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
3   https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
4   https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
..                                                ...  
81  https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
82  https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
83  https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
84  https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  
85  https://www.lcsd.gov.hk/clpss/en/webApp/Fitnes...  

[86 rows x 4 columns]

# Why google search url cannot be worked?

In [28]:
# url = 'https://www.google.com/search?q=japanese+onsen&tbm=isch'
url = 'https://www.google.com/search?q=japanese+onsen'
req = urllib.request.urlopen(url)

In [29]:
req.read()

b'<!DOCTYPE html><html lang="zh-HK"><head><title>Google Search</title><style>body{background-color:#fff}</style><script nonce="PQgWPBiIOkaoE7xKKSPnJA">window.google = window.google || {};window.google.c = window.google.c || {cap:0};</script></head><body><noscript><style>table,div,span,p{display:none}</style><meta content="0;url=/httpservice/retry/enablejs?sei=Bq3naN34HpKl0PEPz8_gqQM" http-equiv="refresh"><div style="display:block">&#33509;&#24744;&#22312;&#25976;&#31186;&#20839;&#20173;&#26410;&#33021;&#33258;&#21205;&#36339;&#36681;&#65292;&#35531;&#40670;&#25802;<a href="/httpservice/retry/enablejs?sei=Bq3naN34HpKl0PEPz8_gqQM">&#36889;&#35023;</a>&#12290;</div></noscript><script nonce="PQgWPBiIOkaoE7xKKSPnJA">(function(){var sctm=false;(function(){sctm&&google.tick("load","pbsst");}).call(this);})();</script><script nonce="PQgWPBiIOkaoE7xKKSPnJA">//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjogMywic291cmNlcyI6WyIiXSwic291cmNlc0NvbnRlbnQiOlsiICJdLCJuYW1

Also, some of sites do not work...

In [ ]:
url = 'https://www.skyscanner.com.hk/transport/flights/hkg/lond/?null&oym=2512&selectedoday=21&iym=2512&selectediday=31'
req = urllib.request.urlopen(url)
html = req.read()

In [ ]:
print(BeautifulSoup(html, 'html.parser').prettify())

tag list: https://www.w3schools.com/tags/ref_byfunc.asp